kpi/management/commands/remove_duplicate_assetversions.py

```diff
@@ -0,0 +1,208 @@
+import json
+from collections import defaultdict
+from hashlib import md5
+from optparse import make_option
+
+from django.contrib.auth.models import User
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from ...models import Asset, AssetVersion
+
+ROUGH_BATCH_MEM_LIMIT_MB = 100
+MAX_BATCH_SIZE = 100
+
+def find_original_and_duplicate_versions(version_pks, asset_pk):
+    '''
+    Given a list of `AssetVersion` primary keys, returns a tuple of:
+        * a list of the original `AssetVersion` primary keys;
+        * a list of the duplicate primary keys;
+        * the batch size used to fetch the versions without memory exhaustion.
+    Duplicates are identified by the following method:
+        * Remove all `$kuid`s from `version_content['survey']` and
+            `version_content['choices']`;
+        * Serialize the modified `version_content`, `deployed_content`, `name`,
+            `_deployment_data`, and `deployed` to JSON;
+        * Calculate the MD5 digest of that JSON;
+        * Consider the first `AssetVersion` (ordered by `pk`) with a given MD5
+            to be the original, and any subsequent `AssetVersion`s with the
+            same MD5 to be duplicates.
+
+    :param version_pks: an iterable of `AssetVersion` primary keys to search
+        for duplicates. They MUST all belong to the same `Asset`.
+    :param asset_pk: the primary key of the `Asset` to which all versions
+        belong. This is required as a safety check.
+    '''
+    version_pks = sorted(version_pks)
+    digests_to_first_version_pks = {}
+    duplicate_version_pks = []
+
+    start = 0
+    batch_size = 1
+    batch_size_guessed = False
+
+    while True:
+        this_batch_version_pks = version_pks[start:start + batch_size]
+        if not this_batch_version_pks:
+            break
+        versions = AssetVersion.objects.filter(
+            asset_id=asset_pk,
+            pk__in=this_batch_version_pks
+        ).order_by('pk')
+        for version in versions.iterator():
+            for kuid_containing in 'survey', 'choices':
+                try:
+                    for item in version.version_content[kuid_containing]:
+                        try:
+                            del item['$kuid']
+                        except KeyError:
+                            pass
+                except KeyError:
+                    continue
+            serialized = json.dumps((
+                version.deployed_content,
+                version.name,
+                version._deployment_data,
+                version.version_content,
+                version.deployed
+            ))
```

Comment:
`json.dumps([...], sort_keys=True)` will fix the issue where md5 hashes of equivalent dicts:

* `{"type": "text", "name": "abc"}`
* `{"name": "abc", "type": "text"}`

would not match. (also available in `formpack.utils.json_hash`)